File: jsimple-util/c#/jsimple/util/Base64.cs
using System;

namespace jsimple.util
{


	/// <summary>
	/// Provides Base64 encoding and decoding as defined by RFC 2045. <p/> <para>This class implements section <cite>6.8. Base64
	/// Content-Transfer-Encoding</cite> from RFC 2045 <cite>Multipurpose Internet Mail Extensions (MIME) Part One: Format of
	/// Internet Message Bodies</cite> by Freed and Borenstein.</para>
	/// <p/>
	/// Sent to us from Nurmi Raimo.T (Nokia-MP/Espoo) <raimo.t.nurmi@nokia.com> & Lindholm Samu (Nokia-MP/Espoo)
	/// <Samu.Lindholm@nokia.com>.  It's original package name was "mango.client.util".
	/// 
	/// @author Apache Software Foundation
	/// @version $Id: Base64.java,v 1.20 2004/05/24 00:21:24 ggregory Exp $ </summary>
	/// <seealso cref= <a href="http://www.ietf.org/rfc/rfc2045.txt">RFC 2045</a>
	/// @since 1.0-dev </seealso>
	public class Base64
	{
		/// <summary>
		/// Chunk size per RFC 2045 section 6.8. <p/> <para>The {@value} character limit does not count the trailing CRLF, but
		/// counts all other characters, including any equal signs.</para>
		/// </summary>
		/// <seealso cref= <a href="http://www.ietf.org/rfc/rfc2045.txt">RFC 2045 section 6.8</a> </seealso>
		internal const int CHUNK_SIZE = 76;

		/// <summary>
		/// Chunk separator per RFC 2045 section 2.1.
		/// </summary>
		/// <seealso cref= <a href="http://www.ietf.org/rfc/rfc2045.txt">RFC 2045 section 2.1</a> </seealso>
		internal static readonly sbyte[] CHUNK_SEPARATOR = new sbyte[] {0xC, 0xA}; // Formerly "\r\n".getBytes();

		/// <summary>
		/// The base length.
		/// </summary>
		internal const int BASELENGTH = 255;

		/// <summary>
		/// Lookup length.
		/// </summary>
		internal const int LOOKUPLENGTH = 64;

		/// <summary>
		/// Used to calculate the number of bits in a byte.
		/// </summary>
		internal const int EIGHTBIT = 8;

		/// <summary>
		/// Used when encoding something which has fewer than 24 bits.
		/// </summary>
		internal const int SIXTEENBIT = 16;

		/// <summary>
		/// Used to determine how many bits data contains.
		/// </summary>
		internal const int TWENTYFOURBITGROUP = 24;

		/// <summary>
		/// Used to get the number of Quadruples.
		/// </summary>
		internal const int FOURBYTE = 4;

		/// <summary>
		/// Used to test the sign of a byte.
		/// </summary>
		internal const int SIGN = -128;

		/// <summary>
		/// Byte used to pad output.
		/// </summary>
		internal static readonly sbyte PAD = (sbyte) '=';

		// Create arrays to hold the base64 characters and a
		// lookup for base64 chars
		private static sbyte[] base64Alphabet = new sbyte[BASELENGTH];
		private static sbyte[] lookUpBase64Alphabet = new sbyte[LOOKUPLENGTH];

		// Populating the lookup and character arrays
		static Base64()
		{
			for (int i = 0; i < BASELENGTH; i++)
				base64Alphabet[i] = (sbyte) -1;
			for (int i = 'Z'; i >= 'A'; i--)
				base64Alphabet[i] = (sbyte)(i - 'A');
			for (int i = 'z'; i >= 'a'; i--)
				base64Alphabet[i] = (sbyte)(i - 'a' + 26);
			for (int i = '9'; i >= '0'; i--)
				base64Alphabet[i] = (sbyte)(i - '0' + 52);

			base64Alphabet['+'] = 62;
			base64Alphabet['/'] = 63;

			for (int i = 0; i <= 25; i++)
				lookUpBase64Alphabet[i] = (sbyte)('A' + i);

			for (int i = 26, j = 0; i <= 51; i++, j++)
				lookUpBase64Alphabet[i] = (sbyte)('a' + j);

			for (int i = 52, j = 0; i <= 61; i++, j++)
				lookUpBase64Alphabet[i] = (sbyte)('0' + j);

			lookUpBase64Alphabet[62] = (sbyte) '+';
			lookUpBase64Alphabet[63] = (sbyte) '/';
		}

		private static bool isBase64(sbyte octect)
		{
			if (octect == PAD)
				return true;
			else if (base64Alphabet[octect] == -1)
				return false;
			else
				return true;
		}

		/// <summary>
		/// Encodes binary data using the base64 algorithm but does not chunk the output.
		/// </summary>
		/// <param name="binaryData"> binary data to encode </param>
		/// <returns> Base64 characters </returns>
		public static sbyte[] encodeBase64(sbyte[] binaryData)
		{
			return encodeBase64(binaryData, false);
		}

		/// <summary>
		/// Encodes binary data using the base64 algorithm but does not chunk the output.  Returns the Base64 data as a
		/// string (which contains at most 64 unique characters).
		/// </summary>
		/// <param name="binaryData"> binary data to encode </param>
		/// <returns> Base64 characters, as a string </returns>
		public static string encodeBase64AsString(sbyte[] binaryData)
		{
			return StringUtils.toStringFromLatin1Bytes(encodeBase64(binaryData, false));
		}

		/// <summary>
		/// Encodes binary data using the base64 algorithm, optionally chunking the output into 76 character blocks.
		/// </summary>
		/// <param name="binaryData"> Array containing binary data to encode. </param>
		/// <param name="isChunked">  if isChunked is true this encoder will chunk the base64 output into 76 character blocks </param>
		/// <returns> Base64-encoded data. </returns>
		public static sbyte[] encodeBase64(sbyte[] binaryData, bool isChunked)
		{
			int lengthDataBits = binaryData.Length * EIGHTBIT;
			int fewerThan24bits = lengthDataBits % TWENTYFOURBITGROUP;
			int numberTriplets = lengthDataBits / TWENTYFOURBITGROUP;
			sbyte[] encodedData = null;
			int encodedDataLength = 0;
			int nbrChunks = 0;

			if (fewerThan24bits != 0)
				//data not divisible by 24 bit
				encodedDataLength = (numberTriplets + 1) * 4;
			else
				// 16 or 8 bit
				encodedDataLength = numberTriplets * 4;

			// If the output is to be "chunked" into 76 character sections,
			// for compliance with RFC 2045 MIME, then it is important to
			// allow for extra length to account for the separator(s)
			if (isChunked)
			{
				nbrChunks = (CHUNK_SEPARATOR.Length == 0 ? 0 : (int) Math.Ceiling((float) encodedDataLength / CHUNK_SIZE));
				encodedDataLength += nbrChunks * CHUNK_SEPARATOR.Length;
			}

			encodedData = new sbyte[encodedDataLength];

			sbyte k = 0, l = 0, b1 = 0, b2 = 0, b3 = 0;

			int encodedIndex = 0;
			int dataIndex = 0;
			int i = 0;
			int nextSeparatorIndex = CHUNK_SIZE;
			int chunksSoFar = 0;

			for (i = 0; i < numberTriplets; i++)
			{
				dataIndex = i * 3;
				b1 = binaryData[dataIndex];
				b2 = binaryData[dataIndex + 1];
				b3 = binaryData[dataIndex + 2];

				//log.debug("b1= " + b1 +", b2= " + b2 + ", b3= " + b3);

				l = (sbyte)(b2 & 0x0f);
				k = (sbyte)(b1 & 0x03);

				sbyte val1 = ((b1 & SIGN) == 0) ? (sbyte)(b1 >> 2) : (sbyte)((b1) >> 2 ^ 0xc0);
				sbyte val2 = ((b2 & SIGN) == 0) ? (sbyte)(b2 >> 4) : (sbyte)((b2) >> 4 ^ 0xf0);
				sbyte val3 = ((b3 & SIGN) == 0) ? (sbyte)(b3 >> 6) : (sbyte)((b3) >> 6 ^ 0xfc);

				encodedData[encodedIndex] = lookUpBase64Alphabet[val1];
				encodedData[encodedIndex + 1] = lookUpBase64Alphabet[val2 | (k << 4)];
				encodedData[encodedIndex + 2] = lookUpBase64Alphabet[(l << 2) | val3];
				encodedData[encodedIndex + 3] = lookUpBase64Alphabet[b3 & 0x3f];

				encodedIndex += 4;

				// If we are chunking, let's put a chunk separator down.
				if (isChunked)
				{
					// this assumes that CHUNK_SIZE % 4 == 0
					if (encodedIndex == nextSeparatorIndex)
					{
						Array.Copy(CHUNK_SEPARATOR, 0, encodedData, encodedIndex, CHUNK_SEPARATOR.Length);
						chunksSoFar++;
						nextSeparatorIndex = (CHUNK_SIZE * (chunksSoFar + 1)) + (chunksSoFar * CHUNK_SEPARATOR.Length);
						encodedIndex += CHUNK_SEPARATOR.Length;
					}
				}
			}

			// form integral number of 6-bit groups
			dataIndex = i * 3;

			if (fewerThan24bits == EIGHTBIT)
			{
				b1 = binaryData[dataIndex];
				k = (sbyte)(b1 & 0x03);
				sbyte val1 = ((b1 & SIGN) == 0) ? (sbyte)(b1 >> 2) : (sbyte)((b1) >> 2 ^ 0xc0);
				encodedData[encodedIndex] = lookUpBase64Alphabet[val1];
				encodedData[encodedIndex + 1] = lookUpBase64Alphabet[k << 4];
				encodedData[encodedIndex + 2] = PAD;
				encodedData[encodedIndex + 3] = PAD;
			}
			else if (fewerThan24bits == SIXTEENBIT)
			{

				b1 = binaryData[dataIndex];
				b2 = binaryData[dataIndex + 1];
				l = (sbyte)(b2 & 0x0f);
				k = (sbyte)(b1 & 0x03);

				sbyte val1 = ((b1 & SIGN) == 0) ? (sbyte)(b1 >> 2) : (sbyte)((b1) >> 2 ^ 0xc0);
				sbyte val2 = ((b2 & SIGN) == 0) ? (sbyte)(b2 >> 4) : (sbyte)((b2) >> 4 ^ 0xf0);

				encodedData[encodedIndex] = lookUpBase64Alphabet[val1];
				encodedData[encodedIndex + 1] = lookUpBase64Alphabet[val2 | (k << 4)];
				encodedData[encodedIndex + 2] = lookUpBase64Alphabet[l << 2];
				encodedData[encodedIndex + 3] = PAD;
			}

			if (isChunked)
			{
				// we also add a separator to the end of the final chunk.
				if (chunksSoFar < nbrChunks)
					Array.Copy(CHUNK_SEPARATOR, 0, encodedData, encodedDataLength - CHUNK_SEPARATOR.Length, CHUNK_SEPARATOR.Length);
			}

			return encodedData;
		}

		/// <summary>
		/// Decodes Base64 data into octects
		/// </summary>
		/// <param name="base64Data"> Byte array containing Base64 data </param>
		/// <returns> Array containing decoded data. </returns>
		public static sbyte[] decodeBase64(sbyte[] base64Data)
		{
			// RFC 2045 requires that we discard ALL non-Base64 characters
			base64Data = discardNonBase64(base64Data);

			// handle the edge case, so we don't have to worry about it later
			if (base64Data.Length == 0)
				return new sbyte[0];

			int numberQuadruple = base64Data.Length / FOURBYTE;
			sbyte[] decodedData = null;
			sbyte b1 = 0, b2 = 0, b3 = 0, b4 = 0, marker0 = 0, marker1 = 0;

			// Throw away anything not in base64Data

			int encodedIndex = 0;
			int dataIndex = 0;
			{
				// this sizes the output array properly - rlw
				int lastData = base64Data.Length;
				// ignore the '=' padding
				while (base64Data[lastData - 1] == PAD)
				{
					if (--lastData == 0)
						return new sbyte[0];
				}
				decodedData = new sbyte[lastData - numberQuadruple];
			}

			for (int i = 0; i < numberQuadruple; i++)
			{
				dataIndex = i * 4;
				marker0 = base64Data[dataIndex + 2];
				marker1 = base64Data[dataIndex + 3];

				b1 = base64Alphabet[base64Data[dataIndex]];
				b2 = base64Alphabet[base64Data[dataIndex + 1]];

				if (marker0 != PAD && marker1 != PAD)
				{
					//No PAD e.g 3cQl
					b3 = base64Alphabet[marker0];
					b4 = base64Alphabet[marker1];

					decodedData[encodedIndex] = (sbyte)(b1 << 2 | b2 >> 4);
					decodedData[encodedIndex + 1] = (sbyte)(((b2 & 0xf) << 4) | ((b3 >> 2) & 0xf));
					decodedData[encodedIndex + 2] = (sbyte)(b3 << 6 | b4);
				}
				else if (marker0 == PAD)
					//Two PAD e.g. 3c[Pad][Pad]
					decodedData[encodedIndex] = (sbyte)(b1 << 2 | b2 >> 4);
				else if (marker1 == PAD)
				{
					//One PAD e.g. 3cQ[Pad]
					b3 = base64Alphabet[marker0];

					decodedData[encodedIndex] = (sbyte)(b1 << 2 | b2 >> 4);
					decodedData[encodedIndex + 1] = (sbyte)(((b2 & 0xf) << 4) | ((b3 >> 2) & 0xf));
				}
				encodedIndex += 3;
			}
			return decodedData;
		}

		/// <summary>
		/// Discards any characters outside of the base64 alphabet, per the requirements on page 25 of RFC 2045 - "Any
		/// characters outside of the base64 alphabet are to be ignored in base64 encoded data."
		/// </summary>
		/// <param name="data"> The base-64 encoded data to groom </param>
		/// <returns> The data, less non-base64 characters (see RFC 2045). </returns>
		internal static sbyte[] discardNonBase64(sbyte[] data)
		{
			sbyte[] groomedData = new sbyte[data.Length];
			int bytesCopied = 0;

			for (int i = 0; i < data.Length; i++)
			{
				if (isBase64(data[i]))
					groomedData[bytesCopied++] = data[i];
			}

			sbyte[] packedData = new sbyte[bytesCopied];

			Array.Copy(groomedData, 0, packedData, 0, bytesCopied);

			return packedData;
		}
	}

}

File: jsimple-util/src/main/java/jsimple/util/Base64.java
/*
 * TODO: UPDATE COPYRIGHT
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jsimple.util;

/**
 * Provides Base64 encoding and decoding as defined by RFC 2045.
 * <p>
 * <p>
 * This class implements section <cite>6.8. Base64 Content-Transfer-Encoding</cite> from RFC 2045 <cite>Multipurpose
 * Internet Mail Extensions (MIME) Part One: Format of Internet Message Bodies</cite> by Freed and Borenstein.
 * </p>
 * <p>
 * Since this class operates directly on byte streams, and not character streams, it is hard-coded to only encode/decode
 * character encodings which are compatible with the lower 127 ASCII chart (ISO-8859-1, Windows-1252, UTF-8, etc).
 * </p>
 *
 * @author Apache Software Foundation
 * @version $Id: Base64.java 1489533 2013-06-04 17:49:00Z sebb $
 * @see <a href="http://www.ietf.org/rfc/rfc2045.txt">RFC 2045</a>
 * @since 2.2
 */
public class Base64 {
    private static final int DEFAULT_BUFFER_RESIZE_FACTOR = 2;

    private static final int DEFAULT_BUFFER_SIZE = 8192;

    /**
     * This array is a lookup table that translates 6-bit positive integer index values into their "Base64 Alphabet"
     * equivalents as specified in Table 1 of RFC 2045.
     * <p>
     * Thanks to "commons" project in ws.apache.org for this code.
     * http://svn.apache.org/repos/asf/webservices/commons/trunk/modules/util/
     */
    private static final byte[] STANDARD_ENCODE_TABLE = {
            'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
            'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
            'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
            'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
            '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '+', '/'
    };

    /**
     * Byte used to pad output.
     */
    private static final byte PAD = '=';

    /**
     * This array is a lookup table that translates Unicode characters drawn from the "Base64 Alphabet" (as specified in
     * Table 1 of RFC 2045) into their 6-bit positive integer equivalents. Characters that are not in the Base64
     * alphabet but fall within the bounds of the array are translated to -1.
     * <p>
     * Note: '+' and '-' both decode to 62. '/' and '_' both decode to 63. This means decoder seamlessly handles both
     * URL_SAFE and STANDARD base64. (The encoder, on the other hand, needs to know ahead of time what to emit).
     * <p>
     * Thanks to "commons" project in ws.apache.org for this code.
     * http://svn.apache.org/repos/asf/webservices/commons/trunk/modules/util/
     */
    private static final byte[] DECODE_TABLE = {
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
            -1, -1, -1, -1, -1, -1, -1, -1, -1, 62, -1, 62, -1, 63, 52, 53, 54,
            55, 56, 57, 58, 59, 60, 61, -1, -1, -1, -1, -1, -1, -1, 0, 1, 2, 3, 4,
            5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23,
            24, 25, -1, -1, -1, -1, 63, -1, 26, 27, 28, 29, 30, 31, 32, 33, 34,
            35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51
    };

    /**
     * Mask used to extract 6 bits, used when encoding
     */
    private static final int MASK_6BITS = 0x3f;

    /**
     * Mask used to extract 8 bits, used in decoding base64 bytes
     */
    private static final int MASK_8BITS = 0xff;

    // The static final fields above are used for the original static byte[] methods on Base64.
    // The private member fields below are used with the new streaming approach, which requires
    // some state be preserved between calls of encode() and decode().

    /**
     * Convenience constant to help us determine when our buffer is going to run out of room and needs resizing.
     * <code>DECODE_SIZE = 3 + lineSeparator.length;</code>
     */
    private static final int DECODE_SIZE = 3;

    /**
     * Convenience constant to help us determine when our buffer is going to run out of room and needs resizing.
     * <code>ENCODE_SIZE = 4 + lineSeparator.length;</code>
     */
    private static final int ENCODE_SIZE = 4;

    /**
     * Buffer for streaming.
     */
    private byte[] buffer;

    /**
     * Position where next character should be written in the buffer.
     */
    private int pos;

    /**
     * Position where next character should be read from the buffer.
     */
    private int readPos;

    /**
     * Writes to the buffer only occur after every 3 reads when encoding, an every 4 reads when decoding. This variable
     * helps track that.
     */
    private int modulus;

    /**
     * Boolean flag to indicate the EOF has been reached. Once EOF has been reached, this Base64 object becomes useless,
     * and must be thrown away.
     */
    private boolean eof;

    /**
     * Place holder for the 3 bytes we're dealing with for our base64 logic. Bitwise operations store and extract the
     * base64 encoding or decoding from this variable.
     */
    private int x;

    /**
     * Returns whether or not the <code>octet</code> is in the base 64 alphabet.
     *
     * @param octet The value to test
     * @return <code>true</code> if the value is defined in the the base 64 alphabet, <code>false</code> otherwise.
     * @since 1.4
     */
    public static boolean isBase64(byte octet) {
        return octet == PAD || (octet >= 0 && octet < DECODE_TABLE.length && DECODE_TABLE[octet] != -1);
    }

    /**
     * Encodes binary data using the base64 algorithm.
     *
     * @param binaryData binary data to encode
     * @return String containing Base64 characters.
     * @since 3.2
     */
    public static String encodeBase64String(byte[] binaryData) {
        return StringUtils.toStringFromLatin1Bytes(encodeBase64(binaryData));
    }

    /**
     * Encodes binary data using the base64 algorithm.
     *
     * @param binaryData Array containing binary data to encode.
     * @return Base64-encoded data.
     * @throws IllegalArgumentException Thrown when the input array needs an output array bigger than maxResultSize
     * @since 1.4
     */
    public static byte[] encodeBase64(byte[] binaryData) {
        return new Base64().encode(binaryData);
    }

    /**
     * Decodes a Base64 String into octets
     *
     * @param base64String String containing Base64 data
     * @return Array containing decoded data.
     * @since 1.4
     */
    public static byte[] decodeBase64(String base64String) {
        return new Base64().decode(base64String);
    }

    /**
     * Decodes Base64 data into octets
     *
     * @param base64Data Byte array containing Base64 data
     * @return Array containing decoded data.
     */
    public static byte[] decodeBase64(byte[] base64Data) {
        return new Base64().decode(base64Data);
    }

    /**
     * Creates a Base64 codec.
     *
     * @since 1.4
     */
    public Base64() {
    }

    /**
     * Returns the amount of buffered data available for reading.
     *
     * @return The amount of buffered data available for reading.
     */
    int avail() {
        return buffer != null ? pos - readPos : 0;
    }

    /**
     * Doubles our buffer.
     */
    private void resizeBuffer() {
        if (buffer == null) {
            buffer = new byte[DEFAULT_BUFFER_SIZE];
            pos = 0;
            readPos = 0;
        } else {
            byte[] b = new byte[buffer.length * DEFAULT_BUFFER_RESIZE_FACTOR];
            PlatformUtils.copyBytes(buffer, 0, b, 0, buffer.length);
            buffer = b;
        }
    }

    /**
     * Extracts buffered data into the provided byte[] array, starting at position bPos, up to a maximum of bAvail
     * bytes. Returns how many bytes were actually extracted.
     *
     * @param b      byte[] array to extract the buffered data into.
     * @param bPos   position in byte[] array to start extraction at.
     * @param bAvail amount of bytes we're allowed to extract. We may extract fewer (if fewer are available).
     * @return The number of bytes successfully extracted into the provided byte[] array.
     */
    int readResults(byte[] b, int bPos, int bAvail) {
        if (buffer != null) {
            int len = Math.min(avail(), bAvail);
            if (buffer != b) {
                PlatformUtils.copyBytes(buffer, readPos, b, bPos, len);
                readPos += len;
                if (readPos >= pos) {
                    buffer = null;
                }
            } else {
                // Re-using the original consumer's output array is only
                // allowed for one round.
                buffer = null;
            }
            return len;
        }
        return eof ? -1 : 0;
    }

    /**
     * Sets the streaming buffer. This is a small optimization where we try to buffer directly to the consumer's output
     * array for one round (if the consumer calls this method first) instead of starting our own buffer.
     *
     * @param out      byte[] array to buffer directly to.
     * @param outPos   Position to start buffering into.
     * @param outAvail Amount of bytes available for direct buffering.
     */
    void setInitialBuffer(byte[] out, int outPos, int outAvail) {
        // We can re-use consumer's original output array under
        // special circumstances, saving on some System.arraycopy().
        if (out != null && out.length == outAvail) {
            buffer = out;
            pos = outPos;
            readPos = outPos;
        }
    }

    /**
     * <p>
     * Encodes all of the provided data, starting at inPos, for inAvail bytes. Must be called at least twice: once with
     * the data to encode, and once with inAvail set to "-1" to alert encoder that EOF has been reached, so flush last
     * remaining bytes (if not multiple of 3).
     * </p>
     * <p>
     * Thanks to "commons" project in ws.apache.org for the bitwise operations, and general approach.
     * http://svn.apache.org/repos/asf/webservices/commons/trunk/modules/util/
     * </p>
     *
     * @param in      byte[] array of binary data to base64 encode.
     * @param inPos   Position to start reading data from.
     * @param inAvail Amount of bytes available from input for encoding.
     */
    void encode(byte[] in, int inPos, int inAvail) {
        if (eof) {
            return;
        }
        // inAvail < 0 is how we're informed of EOF in the underlying data we're
        // encoding.
        if (inAvail < 0) {
            eof = true;
            if (buffer == null || buffer.length - pos < ENCODE_SIZE) {
                resizeBuffer();
            }
            switch (modulus) {
                case 1:
                    buffer[pos++] = STANDARD_ENCODE_TABLE[(x >> 2) & MASK_6BITS];
                    buffer[pos++] = STANDARD_ENCODE_TABLE[(x << 4) & MASK_6BITS];
                    // URL-SAFE skips the padding to further reduce size.
                    buffer[pos++] = PAD;
                    buffer[pos++] = PAD;
                    break;

                case 2:
                    buffer[pos++] = STANDARD_ENCODE_TABLE[(x >> 10) & MASK_6BITS];
                    buffer[pos++] = STANDARD_ENCODE_TABLE[(x >> 4) & MASK_6BITS];
                    buffer[pos++] = STANDARD_ENCODE_TABLE[(x << 2) & MASK_6BITS];
                    buffer[pos++] = PAD;
                    break;
                default:
                    break;  // other values ignored
            }
        } else {
            for (int i = 0; i < inAvail; i++) {
                if (buffer == null || buffer.length - pos < ENCODE_SIZE) {
                    resizeBuffer();
                }
                modulus = (++modulus) % 3;
                int b = in[inPos++];
                if (b < 0) {
                    b += 256;
                }
                x = (x << 8) + b;
                if (0 == modulus) {
                    buffer[pos++] = STANDARD_ENCODE_TABLE[(x >> 18) & MASK_6BITS];
                    buffer[pos++] = STANDARD_ENCODE_TABLE[(x >> 12) & MASK_6BITS];
                    buffer[pos++] = STANDARD_ENCODE_TABLE[(x >> 6) & MASK_6BITS];
                    buffer[pos++] = STANDARD_ENCODE_TABLE[x & MASK_6BITS];
                }
            }
        }
    }

    /**
     * <p>
     * Decodes all of the provided data, starting at inPos, for inAvail bytes. Should be called at least twice: once
     * with the data to decode, and once with inAvail set to "-1" to alert decoder that EOF has been reached. The "-1"
     * call is not necessary when decoding, but it doesn't hurt, either.
     * </p>
     * <p>
     * Ignores all non-base64 characters. This is how chunked (e.g. 76 character) data is handled, since CR and LF are
     * silently ignored, but has implications for other bytes, too. This method subscribes to the garbage-in,
     * garbage-out philosophy: it will not check the provided data for validity.
     * </p>
     * <p>
     * Thanks to "commons" project in ws.apache.org for the bitwise operations, and general approach.
     * http://svn.apache.org/repos/asf/webservices/commons/trunk/modules/util/
     * </p>
     *
     * @param in      byte[] array of ascii data to base64 decode.
     * @param inPos   Position to start reading data from.
     * @param inAvail Amount of bytes available from input for encoding.
     */
    void decode(byte[] in, int inPos, int inAvail) {
        if (eof) {
            return;
        }
        if (inAvail < 0) {
            eof = true;
        }
        for (int i = 0; i < inAvail; i++) {
            if (buffer == null || buffer.length - pos < DECODE_SIZE) {
                resizeBuffer();
            }
            byte b = in[inPos++];
            if (b == PAD) {
                // We're done.
                eof = true;
                break;
            } else {
                if (b >= 0 && b < DECODE_TABLE.length) {
                    int result = DECODE_TABLE[b];
                    if (result >= 0) {
                        modulus = (++modulus) % 4;
                        x = (x << 6) + result;
                        if (modulus == 0) {
                            buffer[pos++] = (byte) ((x >> 16) & MASK_8BITS);
                            buffer[pos++] = (byte) ((x >> 8) & MASK_8BITS);
                            buffer[pos++] = (byte) (x & MASK_8BITS);
                        }
                    }
                }
            }
        }

        // Two forms of EOF as far as base64 decoder is concerned: actual
        // EOF (-1) and first time '=' character is encountered in stream.
        // This approach makes the '=' padding characters completely optional.
        if (eof && modulus != 0) {
            x = x << 6;
            switch (modulus) {
                case 2:
                    x = x << 6;
                    buffer[pos++] = (byte) ((x >> 16) & MASK_8BITS);
                    break;
                case 3:
                    buffer[pos++] = (byte) ((x >> 16) & MASK_8BITS);
                    buffer[pos++] = (byte) ((x >> 8) & MASK_8BITS);
                    break;
                default:
                    break;  // other values ignored
            }
        }
    }

    /**
     * Decodes a String containing containing characters in the Base64 alphabet.
     *
     * @param pArray A String containing Base64 character data
     * @return a byte array containing binary data
     * @since 1.4
     */
    public byte[] decode(String pArray) {
        return decode(StringUtils.toLatin1BytesFromString(pArray));
    }

    /**
     * Decodes a byte[] containing containing characters in the Base64 alphabet.
     *
     * @param pArray A byte array containing Base64 character data
     * @return a byte array containing binary data
     */
    public byte[] decode(byte[] pArray) {
        reset();
        if (pArray.length == 0) {
            return pArray;
        }

        long len = (pArray.length * 3) / 4;
        byte[] buf = new byte[(int) len];
        setInitialBuffer(buf, 0, buf.length);
        decode(pArray, 0, pArray.length);
        decode(pArray, 0, -1); // Notify decoder of EOF.

        // Would be nice to just return buf (like we sometimes do in the encode
        // logic), but we have no idea what the line-length was (could even be
        // variable).  So we cannot determine ahead of time exactly how big an
        // array is necessary.  Hence the need to construct a 2nd byte array to
        // hold the final result:

        byte[] result = new byte[pos];
        readResults(result, 0, result.length);
        return result;
    }

    /**
     * Encodes a byte[] containing binary data, into a byte[] containing characters in the Base64 alphabet.
     *
     * @param pArray a byte array containing binary data
     * @return A byte array containing only Base64 character data
     */
    public byte[] encode(byte[] pArray) {
        reset();
        if (pArray.length == 0) {
            return pArray;
        }
        long len = getEncodeLength(pArray);
        byte[] buf = new byte[(int) len];
        setInitialBuffer(buf, 0, buf.length);
        encode(pArray, 0, pArray.length);
        encode(pArray, 0, -1); // Notify encoder of EOF.
        // Encoder might have resized, even though it was unnecessary.
        if (buffer != buf) {
            readResults(buf, 0, buf.length);
        }
        return buf;
    }

    /**
     * Pre-calculates the amount of space needed to base64-encode the supplied array.
     *
     * @param pArray byte[] array which will later be encoded
     * @return amount of space needed to encoded the supplied array.  Returns
     * a long since a max-len array will require Integer.MAX_VALUE + 33%.
     */
    private static long getEncodeLength(byte[] pArray) {
        long len = (pArray.length * 4) / 3;
        long mod = len % 4;
        if (mod != 0) {
            len += 4 - mod;
        }

        return len;
    }

    /**
     * Resets this Base64 object to its initial newly constructed state.
     */
    private void reset() {
        buffer = null;
        pos = 0;
        readPos = 0;
        modulus = 0;
        eof = false;
    }
}
